### PR TITLE
[printing] make detyping resilient to "let x : _ := t in"

### LINKED ISSUE #13160: [printing] detyping a term with an untyped letin
```ocaml
  | BLetIn ->
      let c = detype d { flags with flg_isgoal = false } avoid env sigma (Option.get body) in
      (* Heuristic: we display the type if in Prop *)
      let s = try Retyping.get_sort_family_of (snd env) sigma ty with _ when !Flags.in_debugger || !Flags.in_toplevel -> InType (* Can fail because of sigma missing in debugger *) in
      let t = if s != InProp  && not !Flags.raw_print then None else Some (detype d { flags with flg_isgoal = false } avoid env sigma ty) in
      GLetIn (na', c, t, r)
```
Here `Retyping.get_sort_family_of`, if the input is an evar, raises an error which is only (abruptly) catched if we are in coqtop.
I think we should catch all retyping errors (and only these) unconditionally.

I hit this in Elpi where I want to detype a term which I did not pretype yet.

### PR DESCRIPTION
**Pull request by gares**
Fix #13160

### CODE REVIEW

File: pretyping/detyping.ml [hunk 1]
```diff
@@ -883,7 +883,9 @@ and detype_binder d flags bk avoid env sigma decl c =
   | BLetIn ->
       let c = detype d { flags with flg_isgoal = false } avoid env sigma (Option.get body) in
       (* Heuristic: we display the type if in Prop *)
-      let s = try Retyping.get_sort_family_of (snd env) sigma ty with _ when !Flags.in_debugger || !Flags.in_toplevel -> InType (* Can fail because of sigma missing in debugger *) in
+      let s =
+        try Retyping.get_sort_family_of (snd env) sigma ty
+        with Retyping.RetypeError _ -> InType (* eg: the type is an evar *) in
```
Comment (SkySkimmer) [inline pretyping/detyping.ml:888]:
```suggestion
        with Retyping.RetypeError _ -> InType (* eg: the type is an evar *) 
      in
```

Comment (gares) [reply pretyping/detyping.ml:888]: I'm not a big fan of this style, a line with just a `in` does not seem worth it. Is there a reason for it?

Comment (SkySkimmer) [reply pretyping/detyping.ml:888]: It's much easier to find the end of the letin that way imo

Comment (gares) [reply pretyping/detyping.ml:888]: Well, indentation helps as well no?

Some styles help patches to be smaller, changes more local... I'm all for that.
This one looks a bit too much about taste to me (even re-running CI looks too much compared to what we could gain, imo).

Comment (herbelin) [reply pretyping/detyping.ml:888]: @gares: please keep the information that it can fail also in the debugger or in the OCaml toplevel.

Comment (gares) [reply pretyping/detyping.ml:888]: I thought about that, but then I thought it was outdated, from a time where sigma was not passed explicitly. Now that we pass it, what is the difference when we run it from the debugger? If I'm wrong I'll add it back, sure.

Comment (herbelin) [reply pretyping/detyping.ml:888]: The debugging printers run in the ocamldebug process and are not aware of the `Global.env` of the process they are debugging. That would be nice if ocamldebug were able to pass other information to the printers than the term to print but I'm not sure it is possible. To be honest I did not check if it is possible but as it is now, it is certainly not the case. As another example, the nametab is unused by the debugging printers.

Comment (gares) [reply pretyping/detyping.ml:888]: ok, I'll restore the comment then, and while I'm at it make @SkySkimmer happy

File: pretyping/detyping.ml [hunk 2]
```diff
@@ -883,7 +883,12 @@ and detype_binder d flags bk avoid env sigma decl c =
   | BLetIn ->
       let c = detype d { flags with flg_isgoal = false } avoid env sigma (Option.get body) in
       (* Heuristic: we display the type if in Prop *)
-      let s = try Retyping.get_sort_family_of (snd env) sigma ty with _ when !Flags.in_debugger || !Flags.in_toplevel -> InType (* Can fail because of sigma missing in debugger *) in
+      let s =
+        (* It can fail if ty is an evar, or if run inside ocamldebug or the
+           OCaml toplevel since their printers don't have access to sigma/env *)
```
Comment (herbelin) [inline pretyping/detyping.ml:888]:
```suggestion
           OCaml toplevel since their printers don't have access to the proper sigma/env *)
```

**Review by herbelin — APPROVED** [review]
Good to me, with an extra wording suggestion if you like it. Assigning.

**Comment by gares** [comment]
This looks good to me. CI failure on Mtac2 seems spurious, re-running it.

**Comment by herbelin** [comment]
@coqbot: merge now

**Pull request merged**